File: leetcode/python/234_palindrome_linked_list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: ListNode | None) -> bool:
        r = self.splitHalf(head)
        r = self.reverse(r)
        l = head
        while r:
            if l.val != r.val:
                return False
            l = l.next
            r = r.next
        return True
    
    def splitHalf(self, head: ListNode | None) -> ListNode | None:
        slow = fast = head
        while fast:
            slow = slow.next
            fast = fast.next.next if fast.next else None
        return slow
    
    def reverse(self, head: ListNode | None) -> ListNode | None:
        prev = None
        while head:
            head.next, prev, head = prev, head, head.next
        return prev

File: leetcode/rust/src/234_palindrome_linked_list.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        let mut nums = Vec::new();
        let mut head = head;
        while let Some(node) = head {
            nums.push(node.val);
            head = node.next;
        }
        let mut i = 0;
        let mut j = nums.len() - 1;
        while i < j {
            if nums[i] != nums[j] {
                return false;
            }
            i += 1;
            j -= 1;
        }
        true
    }
}